### PR TITLE
Switch code editor to Monaco

### PR DESCRIPTION
**Pull request by silverwind**
This switches out CodeMirror for [Monaco](https://microsoft.github.io/monaco-editor/) which is based on the same code base as VS Code and should work pretty similar to it.

It does add a few async chunks, totalling around 10MB to our build. It currently supports around 65 languages and in the default configuration, each language would emit one ugly `[number].js` chunk, so I opted to combine them all into a single 4MB file for now.

CodeMirror is still being used under the hood by SimpleMDE so it can not be removed yet.

Screen Shot 2020-05-10 at 22 28 49
Screen Shot 2020-05-10 at 22 27 26

### CODE REVIEW

**Comment by silverwind** [comment]
TODO: Kick out CodeMirror-specific CSS.

**Comment by silverwind** [comment]
Should be good for review now. I decided to not cut features. The bundles are big without a doubt, but it still loads within around 2 seconds on my machine.

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
This is soooo nice :)

**Comment by silverwind** [comment]
There is a issue with the `highlight.worker.js` not being generated, don't land yet.

**Comment by silverwind** [comment]
Actually it's fine, I was just misinterpreting webpack output. `js/highlight.js` is the worker generated from the chunk in `web_src/js/features/highlight.js` which includes the worker inline. This mechanism did not actually change with the PR.

**Comment by jolheiser** [comment]
When creating a new file from the UI, it seemed to be perpetually "loading..."

**Comment by silverwind** [comment]
Will fix that, did not try creating new files yet.

**Comment by silverwind** [comment]
File creation fixed, it was related to bad error handling which I clean up. Also fixed editorconfig being missing in the template code after POST (for example, when filename already exists).

File: web_src/js/features/editors.js [hunk 1]
```diff
@@ -92,12 +92,13 @@ function getOptions(filenameInput, lineWrapExts) {
   const opts = {theme, wordWrap};
   if (isObject(ec)) {
     opts.detectIndentation = !('indent_style' in ec) || !('indent_size' in ec);
-    opts.indentSize = ('indent_size' in ec) ? Number(ec.indent_size) : undefined;
-    opts.tabSize = ('tab_width' in ec) ? Number(ec.tab_width) : opts.indentSize;
+    if ('indent_size' in ec) opts.indentSize = Number(ec.indent_size);
+    if ('tab_width' in ec) opts.tabSize = Number(ec.tab_width) || opts.indentSize;
+    if ('max_line_length' in ec) opts.rulers = [Number(ec.max_line_length)];
```
Comment (silverwind) [inline web_src/js/features/editors.js]: Note this is [lacking support in editorconfig-core-go](https://github.com/editorconfig/editorconfig-core-go/issues/45).

**Comment by silverwind** [comment]
@jolheiser care to have another look?

**Comment by lunny** [comment]
`minimap` could be an option. For a large file, it could be disabled. For smaller files, it's beatiful!!!

**Comment by silverwind** [comment]
Yes, I suppose we can create a small options UI later, especially for things that can not be controlled via `.editorconfig`. [There are many such options](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.editor.istandaloneeditorconstructionoptions.html).

But first, I think we'd need some kind of mechanism to store user UI settings in the database (maybe just a JSON object stored as string).

**Review by jolheiser — APPROVED** [review]
LGTM

**Comment by lafriks** [comment]
make lg-tm work

File: templates/base/head.tmpl [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html lang="{{.Language}}">
+<html lang="{{.Language}}" class="theme-{{.SignedUser.Theme}}">
```
Comment (lunny) [inline templates/base/head.tmpl:2]: How to handle non-login user?

Comment (lafriks) [reply templates/base/head.tmpl:2]: I don't think anonymous users have access to code edditor

Comment (silverwind) [reply templates/base/head.tmpl:2]: he's right, condition should be improved as even unsigned users should have a theme, just not in this variable. Right now it does not matter but I want to move the theme to CSS variables using this class so it has to work then.

**Pull request merged**